server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/join/VerticallyConcatOperator.java

```diff
@@ -0,0 +1,206 @@
+
+  private final OperatorContext operatorContext;
+
+  private final List<Operator> children;
+
+  private final int inputOperatorsCount;
+
+  /** TsBlock from child operator. Only one cache now. */
+  private final TsBlock[] inputTsBlocks;
+
+  /** start index for each input TsBlocks and size of it is equal to inputTsBlocks */
+  private final int[] inputIndex;
+
+  private final int outputColumnCount;
+
+  private final TsBlockBuilder tsBlockBuilder;
+
+  private boolean finished;
+
+  public VerticallyConcatOperator(
+      OperatorContext operatorContext, List<Operator> children, List<TSDataType> dataTypes) {
+    checkArgument(
+        children != null && children.size() > 0,
+        "child size of VerticallyConcatOperator should be larger than 0");
+    this.operatorContext = operatorContext;
+    this.children = children;
+    this.inputOperatorsCount = children.size();
+    this.inputTsBlocks = new TsBlock[this.inputOperatorsCount];
+    this.inputIndex = new int[this.inputOperatorsCount];
+    this.outputColumnCount = dataTypes.size();
+    this.tsBlockBuilder = new TsBlockBuilder(dataTypes);
+  }
+
+  @Override
+  public OperatorContext getOperatorContext() {
+    return operatorContext;
+  }
+
+  @Override
+  public ListenableFuture<?> isBlocked() {
+    List<ListenableFuture<?>> listenableFutures = new ArrayList<>();
+    for (int i = 0; i < inputOperatorsCount; i++) {
+      if (empty(i)) {
+        ListenableFuture<?> blocked = children.get(i).isBlocked();
+        if (!blocked.isDone()) {
+          listenableFutures.add(blocked);
+        }
+      }
+    }
+    return listenableFutures.isEmpty() ? NOT_BLOCKED : successfulAsList(listenableFutures);
+  }
+
+  @Override
+  public TsBlock next() {
+    tsBlockBuilder.reset();
+    // indicates how many rows can be built in this calculate
+    int maxRowCanBuild = Integer.MAX_VALUE;
+    for (int i = 0; i < inputOperatorsCount; i++) {
+      if (empty(i)) {
+        inputIndex[i] = 0;
+        inputTsBlocks[i] = children.get(i).next();
+        if (empty(i)) {
+          // child operator has not prepared TsBlock well
+          return null;
+        }
+      }
+      maxRowCanBuild =
+          Math.min(maxRowCanBuild, inputTsBlocks[i].getPositionCount() - inputIndex[i]);
+    }
+
+    TimeColumn firstTimeColumn = inputTsBlocks[0].getTimeColumn();
+    TimeColumnBuilder timeColumnBuilder = tsBlockBuilder.getTimeColumnBuilder();
+    ColumnBuilder[] valueColumnBuilders = tsBlockBuilder.getValueColumnBuilders();
+
+    for (int row = 0; row < maxRowCanBuild; row++) {
+      // build TimeColumn
+      timeColumnBuilder.writeLong(firstTimeColumn.getLong(inputIndex[0]));
+      tsBlockBuilder.declarePosition();
+    }
+
+    // build each ValueColumn in every inputTsBlock
+    int valueBuilderIndex = 0; // indicate which valueColumnBuilder should use
+    for (int i = 0; i < inputOperatorsCount; i++) {
+      int currTsBlockIndex = inputIndex[i];
```

Comment:
Actually, I didn't see  that you increase this `currTsBlockIndex`, it seems that you always append the same value in the for-loop

Reply:
If so, add a new UT for this Operator.